src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs

```diff
@@ -930,16 +980,54 @@ public ArrayShapeEncoder(BlobBuilder builder)
             Builder = builder;
         }
 
+        /// <summary>
+        /// Encodes array shape.
+        /// </summary>
+        /// <param name="rank">The number of dimensions in the array (shall be 1 or more).</param>
+        /// <param name="sizes">
+        /// Dimension sizes. The array may be shorter than <paramref name="rank"/> but not longer.
+        /// </param>
+        /// <param name="lowerBounds">
+        /// Dimension lower bounds, or <c>default(<see cref="ImmutableArray{Int32}"/>)</c> to set all <paramref name="rank"/> lower bounds to 0. 
+        /// The array may be shorter than <paramref name="rank"/> but not longer.
+        /// </param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="rank"/> is outside of range [1, 0xffff],
+        /// smaller than <paramref name="sizes"/>.Length, or
+        /// smaller than <paramref name="lowerBounds"/>.Length.
+        /// </exception>
+        /// <exception cref="ArgumentNullException"><paramref name="sizes"/> is null.</exception>
         public void Shape(int rank, ImmutableArray<int> sizes, ImmutableArray<int> lowerBounds)
         {
+            // The specification doesn't impose a limit on the max number of array dimensions. 
+            // The CLR supports <64. More than 0xffff is causing crashes in various tools (ildasm).
+            if (unchecked((uint)(rank - 1)) > ushort.MaxValue - 1)
+            {
+                Throw.ArgumentOutOfRange(nameof(rank));
+            }
+
+            if (sizes.IsDefault)
+            {
+                Throw.ArgumentNull(nameof(sizes));
+            }
+
+            // rank
             Builder.WriteCompressedInteger(rank);
+
+            // sizes
+            if (sizes.Length > rank)
+            {
+                Throw.ArgumentOutOfRange(nameof(rank));
+            }
+
             Builder.WriteCompressedInteger(sizes.Length);
             foreach (int size in sizes)
             {
                 Builder.WriteCompressedInteger(size);
             }
 
-            if (lowerBounds.IsDefault)
+            // lower bounds
+            if (lowerBounds.IsDefault) // TODO: remove -- update Roslyn
```

Comment:
I'm tracking it in my list of work to do on SRM this sprint to reduce overhead. Will convert any TODO's remaining after the sprint to git issues.